### ISSUE TITLE
In SignalR, allow strongly typed hubs to return `void` not `Task`

### ISSUE THREAD

**Issue by heroboy** [dotnet/aspnetcore#49717]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Currently, the interface' methods in `Hub<ISomeInterface>` must return `Task`. I want it allow return `void`.
1. I don't care the return value from client.
2. I don't care and want to process the send error.
4. I may not have to write `_ = hub.Clients.Caller.foo()` to explicit ignore the `Task`.
5. If it is a `Task`, let me worried about somewhere there maybe a unhandled exception when send failed.
6. Some other api like `ApiController` allow `void`(or some type other then Task) return method.

### Describe the solution you'd like

This is the example in https://learn.microsoft.com/en-us/aspnet/core/signalr/hubs?view=aspnetcore-8.0
```C#
public interface IChatClient
{
    Task ReceiveMessage(string user, string message);
}
public class StronglyTypedChatHub : Hub<IChatClient>
{
    public async Task SendMessage(string user, string message)
        => await Clients.All.ReceiveMessage(user, message);

    public async Task SendMessageToCaller(string user, string message)
        => await Clients.Caller.ReceiveMessage(user, message);

    public async Task SendMessageToGroup(string user, string message)
        => await Clients.Group("SignalR Users").ReceiveMessage(user, message);
}
```

I want it can be written to
```c#
public interface IChatClient
{
    void ReceiveMessage(string user, string message);
}

public class StronglyTypedChatHub : Hub<IChatClient>
{
    public void SendMessage(string user, string message)
        => Clients.All.ReceiveMessage(user, message);

    public void SendMessageToCaller(string user, string message)
        => Clients.Caller.ReceiveMessage(user, message);

    public void SendMessageToGroup(string user, string message)
        => Clients.Group("SignalR Users").ReceiveMessage(user, message);
}
```

**Comment by heroboy** [comment]
Off topic. 
This issue(https://github.com/dotnet/aspnetcore/issues/47189)  is locked.
Can you mark it as bug. This bug is so cirtical that I always search it to sure what it happened.

**Comment by gfoidl** [comment]
SignalR communication is going to be async, so `Task` makes sense.
Please search for `async void` and why it's [bad](https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#async-void)", then you'll see that this proposal won't be taken. Especially

> I don't care and want to process the send error.

but with `async void` it may crash the process, which is likely more evil than using `Task`.

**Comment by heroboy** [comment]
I think processing send error is meaningless. 
Maybe SignalR has may backend like long polling. I will say some my points assume SignalR is over tcp/websocket.

What you can do when some message is sent failed? Try again? Any send failed will cause disconnect. So try again is no need.
I have nothing to do when it send failed. Just let the client connect again, and send all the data again.

For example, you want to send 1,2,3 to a client. The client may receive 1,2 and disconnect. Will never receive 1, (2 is sent failed) ,3.

I want know what you will do if you need send a static list. 
```csharp
static List<int> list;
async Task foo()
{
    int[] cloned;
    //The real world problem will not be simple as just clone the data and send.
    //
    lock(list) {
          cloned = list.ToArray();
    }
    foreach(var i in cloned) await client.Send(i); 
   //or Task.WaitAll(  cloned.Select(i => client.Send(i)) )
}
```
> SignalR communication is going to be async, so Task makes sense.

When I say 'send', I mean push the data to the send queue. And let low level code to send the data to the client asynchronously. Then "push the data " is not async. If the data is actually failed to sent, just tell me it is disconnected.

I have many words to say, but my English is bad. I just have two points:
1. Nothing I can do when send failed.
2. It is hard to await when the send logic and send data must be locked.

**Comment by heroboy** [comment]
Even the OS has its sending buffer. 
SignalR doesn't need a queue. The underlying transporter has it. Look like in SignalR, it doesn't await before it send the data to the transporter. So the code
```csharp
SendAsync("A");
SendAsync("B");
SendAsync("C");
```
Even the tasks are not be awaited, they run parallel. The client will receviec them in the call order.

And you say crash many times. I have tried many ways to make such crash.
`SendAsync` never throw any exception. Even an invalid ConnectionId doesn't makes it throw. Of cause, `InvokeAsync` will always throw an exception when something wrong. But if I ignore the task, it doesn't make the process crash. And even a warning log is not exist.
But I doesn't try to send large data that makes the transporter layer meet its high water mark. I don't know what will happen.

**Comment by BrennanConroy** [comment]
**Tldr;** As a framework we aren't going to make it easier for you to write code that we think is an anti-pattern. And if you think you know better, there is always `return Task.CompletedTask;`

-----------

Not awaiting asynchronous methods is a bad idea. Just because you "don't care about errors" doesn't mean you should ignore the task.

Let's take a semi-arbitrary example that shows a little bit about the internals of SignalR and why you should `await` the task from `Clients.All.ReceiveMessage(user, message)`

```csharp
var pipe = new Pipe(new PipeOptions(pauseWriterThreshold: 1000, resumeWriterThreshold: 500));

for (var i = 0; i < 100; ++i)
{
    var span = pipe.Writer.GetSpan(1000);

    pipe.Writer.Advance(1000);

    _ = pipe.Writer.FlushAsync();
}
```

Here we don't `await` the `FlushAsync` call on the `Pipe` (which is what SignalR is using internally). So what happens is that our limit of 1000 bytes on the `Pipe` is ignored and we end up writing (allocating) 100kb on the server instead of allocating 1000, waiting for it to be sent over the network, repeat.

If you have enough client calls you could easily allow clients to allocate arbitrary amounts on the server and cause it to OOM (out of memory). This can be avoided by awaiting the tasks which allows backpressure to occur (the tasks force your code to wait and not allocate anymore since they are waiting for the server to flush the bytes to the network).

Another other fun issue if you don't `await`:
```csharp
public void HubMethod()
{
    var item = new Item() { Value = "foo" };
    _ = Clients.All.SendAsync("MethodName", item);

    item.Value = "bar";
    _ = Clients.All.SendAsync("MethodName", item);
}

class Item
{
    public string Value { get; set; }
}
```

In this example the first `SendAsync` call might send `Value = foo` or `Value = bar`. You have no guarantee because you didn't `await` the send call so it might still be processing the item while your code is modifying it.

**Comment by heroboy** [comment]
@BrennanConroy Thank you for your detailed explanation.
Actually I have already aware the two issues you pointed.
1. If it is neccessary, I can collect the Task in the lock statement block, and await them all.
2. This is why I ask question, https://stackoverflow.com/questions/76091923/how-to-lock-a-shared-object-before-serialization-in-asp-net-core-signalr . My current solution is create new object for every sending. This is to solve the race condition of internal data, and separte the internal data structure and protocol data structure.

Thanks again let me sure not await send is very harmful.
